Add aggregation to slotted runtime

systay opened the pull request.

Commit:
Introduce DistinctQueryHorizon and logical plan

Commit:
Make aggregation handle it's own projection

pontusmelke committed:
Added register allocation for aggregation

pontusmelke committed:
Add EagerAggregationRegisterPipe

Commit:
Remove warnings

Commit:
Mark tests as passing in slotted runtime

pontusmelke committed:
Add DistinctRegisterPipe

pontusmelke committed:
Removed tests from blacklist

pontusmelke committed:
Fixed failing tests

pontusmelke committed:
Updated compiled runtime to cope with changes in DISTINCT and AGGREGATION

Commit:
Fix so aggregation grouping gets unapplied as expected

- Also turn off pattern expressions for the slotted runtime

Commit:
Make sure OptionalMatchRemover works with the new QueryGraph

Commit:
Added register allocation for DISTINCT

Commit:
Code style

Commit:
More tests turning green after rebase

Commit:
Simpler (faster) code when aggregating and not grouping

henriknyman committed:
Removed passing tests from blacklist

henriknyman commented on community/cypher/ir-3.3/src/main/scala/org/neo4j/cypher/internal/ir/v3_3/QueryHorizon.scala:
```diff
@@ -113,29 +113,44 @@ final case class RegularQueryProjection(projections: Map[String, Expression] = M
+  override def exposedSymbols(coveredIds: Set[IdName]): Set[IdName] = (groupingExpressions.keys ++  aggregationExpressions.keys).map(IdName.apply).toSet
+}
+
+final case class DistinctQueryProjection(groupingKeys: Map[String, Expression] = Map.empty,
```
I guess this could also be renamed `groupingExpressions` for consistency with the renamings above?

henriknyman commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala:
```diff
@@ -241,6 +237,16 @@ object RegisterAllocation {
+    groupingExpressions foreach {
+      case (key, parserAst.Variable(ident)) =>
+        val slotInfo = incoming(ident)
+        outgoing.newReference(key, slotInfo.nullable, slotInfo.typ)
```
Can't grouping expression also be a primitive Long type?

henriknyman commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/compiled/codegen/ir/aggregation/AggregationConverter.scala:
```diff
@@ -30,7 +30,7 @@ import org.neo4j.cypher.internal.frontend.v3_3.ast
 object AggregationConverter {
 
-  def aggregateExpressionConverter(opName: String, groupingVariables: Iterable[Variable], name: String, e: ast.Expression) (implicit context: CodeGenContext): AggregateExpression = {
+  def aggregateExpressionConverter(opName: String, groupingVariables: Iterable[(String,CodeGenExpression)], name: String, e: ast.Expression) (implicit context: CodeGenContext): AggregateExpression = {
```
A named type (with named fields) for `(String,CodeGenExpression)` would be really nice for readability of the code below.

henriknyman commented on enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/compiled/codegen/ir/aggregation/DynamicCount.scala:
```diff
@@ -47,6 +50,13 @@ class DynamicCount(opName: String, variable: Variable, expression: CodeGenExpres
+      case (v, expr) =>
+        structure.declare(v, expr.codeGenType)
+        if (expr.codeGenType == CodeGenType.Any) structure.assign(v, expr.codeGenType,
+                                                                  structure.materializeAny(expr.generateExpression(structure)))
```
Do we need to materialize here?

henriknyman committed:
Fix bug in ProjectionRegisterPipe

Projection in the slotted runtime did not correctly distinguish between
LongSlot and RefSlot values.

henriknyman committed:
Fix bug in OptionalExpandIntoRegisterPipe

OptionalExpandInto should not return null for any node, since they
are already bound.

henriknyman committed:
Improve debug support for PrimitiveExecutionContext

henriknyman committed:
Remove passing tests from blacklist

henriknyman committed:
Fix failing unit test
